### PR TITLE
[receiver/mongodb] Add Scaffolding for MongoDB receiver 

### PR DESCRIPTION
**Pull request by schmikei**
Sponsor: @djaglowski 

**Description:** Adds scaffolding of a mongodb receiver initially drafted in  #6951, in order to make reviewing easier.

This adds the scaffolding for a new component that will be used for standalone, replica set, and sharded deployments of MongoDB using the [mongo-go driver](https://github.com/mongodb/mongo-go-driver)

Config Proposed:

```yaml
receivers:
  mongodb:
    hosts:
        - endpoint: localhost:27017
    username: otelu
    password: otelp
    collection_interval: 10s
    tls:
      insecure: true
      insecure_skip_verify: true
```

Admittedly the thing I am most torn on is whether to support a list based host approach or just a single endpoint. A list based approach appears the most flexible in my opinion. But this is a point I'd like to get other opinions. 

Expecting to follow up this PR with metrics collection that was present in #6951  

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

Minor unit tests. Verified mongo driver works well on these versions of MongoDB:
- 2.6
- 3.0+
- 4.0+
- 5.0

**Documentation:** <Describe the documentation added.>

Minor documentation on the setup of a clustermonitor user and purpose of this receiver.

### CODE REVIEW

**Comment by bogdandrutu** [comment]
Sorry for my ignorance, but what is the diff with https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/mongodbatlasreceiver?

**Comment by schmikei** [comment]
@bogdandrutu No Problem! [Mongodb Atlas](https://www.mongodb.com/cloud) is a cloud based service that exposes endpoints which that receiver is using to scrape for data. 

Standalone Mongodb's can be installed/managed without atlas as a standalone solution. However this does not give the ability to reach out to the [endpoints that atlas exposes](https://docs.atlas.mongodb.com/reference/api/monitoring-and-logs/). So there must be another mechanism to monitor mongodb metrics that are in this "group" of solutions where mongodb is managed on-prem.

Generally these docs I find to be helpful to navigate against all the different types of mongodb server installations
https://docs.mongodb.com/manual/?_ga=2.157913022.506964807.1641233545-1790256795.1636558386#what-is-mongodb-

**Comment by jpkrohling** [comment]
@schmikei, could you please go over #6926 and make sure this here has everything included in the "first PR" section?

File: .github/CODEOWNERS
Comment (jpkrohling) [reply .github/CODEOWNERS]: This is something we should definitely discuss during the SIG meeting. I would expect the original contributors to be code owners, as they should be responsible for handling bugs.

File: receiver/mongodbreceiver/README.md
Comment (djaglowski) [inline receiver/mongodbreceiver/README.md:24]: Since it has a default, this isn't required.

File: receiver/mongodbreceiver/README.md
Comment (djaglowski) [inline receiver/mongodbreceiver/README.md:26]:
```suggestion
  - For replica sets specify the hostnames and ports of the mongod instances that are in the replica set configuration. If the `replica_set` field is specified, nodes will be autodiscovered.
```

Can you reword this? Maybe something like this?

Comment (djaglowski) [inline receiver/mongodbreceiver/README.md:26]: If `replica_set` is specified, does the user only need to specify one mongod, or is it necessary to specify all of them?

File: receiver/mongodbreceiver/config.go
Comment (djaglowski) [inline receiver/mongodbreceiver/config.go:43]: Comments here are redundant w/ documentation

File: receiver/mongodbreceiver/README.md
Comment (djaglowski) [inline receiver/mongodbreceiver/README.md:34]: What happens if this is specified when the deployment is not a replica set?

File: receiver/mongodbreceiver/config.go
Comment (djaglowski) [inline receiver/mongodbreceiver/config.go:58]:
```suggestion
		_, _, hostPortErr := net.SplitHostPort(host.Endpoint)
		if hostPortErr != nil && host.Transport != "unix" {
```

No need for this extra function and conversion to boolean. It's more concise and idiomatic to just use the error directly.

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:1]:
```suggestion
name: mongodbreceiver

```
Newline is always required.

File: receiver/mongodbreceiver/config.go
Comment (djaglowski) [inline receiver/mongodbreceiver/config.go:56]: Does MongoDB have a default port? If so, shouldn't we allow that to be missing and assume the default when it is?

File: .github/CODEOWNERS
Comment (bogdandrutu) [reply .github/CODEOWNERS]: The name can be added, it will not have any effect, but at least we "manually" track the approval, which we do anyway for the moment.

Adding them to the org is a natural next step after the component is merged.

File: receiver/mongodbreceiver/config.go
Comment (schmikei) [reply receiver/mongodbreceiver/config.go:56]: Yes the default port is actually 27017, I realized you can omit the port portion of the connection URI we are trying to build and it will default to the correct number. I'll revisit this, good catch 👍 

https://docs.mongodb.com/manual/reference/connection-string/#connections-connection-options

File: receiver/mongodbreceiver/README.md
Comment (schmikei) [reply receiver/mongodbreceiver/README.md:26]: From my understanding, it is expected to specify the mongod instances which are specified in the configuration of the replica set. 

https://docs.mongodb.com/manual/reference/connection-string/#replica-set-option

File: receiver/mongodbreceiver/README.md
Comment (schmikei) [reply receiver/mongodbreceiver/README.md:31]: Did a little more digging and the mongo driver won't know that its not a replica set and will fail to connect. With what I have currently spiked out it would give this error:
```
Error: cannot start receivers: unable to validate connection: unable to connect: unable to instantiate mongo client: could not connect to hosts [localhost:27017]: server selection error: server selection timeout, current topology: { Type: ReplicaSetNoPrimary, Servers: [] }
```

File: receiver/mongodbreceiver/config.go
Comment (djaglowski) [inline receiver/mongodbreceiver/config.go:46]: Since we're not doing anything when it is unix, just continue. This keeps the code flatter, which is encouraged in golang.
```suggestion
		if host.Transport == "unix" {
			continue
		}
		
		// continue processing as non-unix
```

File: receiver/mongodbreceiver/config.go
Comment (djaglowski) [inline receiver/mongodbreceiver/config.go:48]: This depends on network interaction, which I don't think we want to do in config validation. Config validation should depend only on the configuration.

File: receiver/mongodbreceiver/config_test.go
Comment (djaglowski) [inline receiver/mongodbreceiver/config_test.go:24]: Let's not use this. If you need `assert` behavior rather than `require`, please use `github.com/stretchr/testify/assert`. In this case, I don't think we need it though.

File: cmd/configschema/go.mod
Comment (djaglowski) [inline cmd/configschema/go.mod:650]: Need to first add this as a dependency above.

**Review by djaglowski — COMMENTED** [review]
Tracking based on the list as it currently stands in #6972:
- [x] Mention the sponsor in the PR description
- [x] A complete config.go, which will show how your component will look like to users
- [x] README.md with some basic information about your component. It doesn't have to be complete, but should contain at least:
  - [x] Purpose and use-cases
  - [x] Telemetry data types supported
- [x] versions.yaml
- [x] CODEOWNERS with at least the sponsor and the component's author
- [ ] internal/components tests
- [ ] go.mod at the root of the repository and cmd/configschema
dependabot
- [x] The skeleton of the component, potentially using the helpers
- [x] Makefile

You still need to the following:
- Explain the use case in the Readme. (Can be as simple as stating that this allows one to monitor metrics from a MongoDB cluster)
- Add to internal/components tests
- Add to root go.mod

File: .github/CODEOWNERS
Comment (djaglowski) [reply .github/CODEOWNERS]: @schmikei, based on this conversation, you should re-add your handle here.

**Comment by djaglowski** [comment]
@schmikei I apologize for the churn, but these two checklist items are a bit premature and should be backed out. (See discussion [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7046#issuecomment-1006862731))
* internal/components tests
* go.mod at the root of the repository and cmd/configschema
dependabot

File: receiver/mongodbreceiver/config.go
Comment (djaglowski) [inline receiver/mongodbreceiver/config.go:35]: This is not described in the docs. 

Am I correct in thinking it will used by the mongo client library?

File: receiver/mongodbreceiver/factory.go
Comment (djaglowski) [inline receiver/mongodbreceiver/factory.go:50]:
```suggestion
			CollectionInterval: time.Minute,
		},
		Timeout: time.Minute,
```

File: receiver/mongodbreceiver/config.go
Comment (schmikei) [reply receiver/mongodbreceiver/config.go:35]: Initial plan was to use it for 

https://github.com/mongodb/mongo-go-driver/blob/384560d7efa96ed08030bf1eaebba0230776b82f/mongo/options/clientoptions.go#L694-L700

and the corresponding socket option in the client.

**Review by djaglowski — APPROVED** [review]

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**
